### PR TITLE
Assignment 3 complete

### PR DESCRIPTION
**Pull request by dwalkes**

### CODE REVIEW

File: examples/systemcalls/systemcalls.c [hunk 1]
```diff
@@ -7,15 +14,22 @@
...
+    int status = system(cmd);
+
+    if(-1 == status){
+	 //Log error
+	  return false;
+    }
+	
+    //What is this???
+    if(!WIFEXITED(status)){
+	    return false;
```
Comment (dwalkes) [inline examples/systemcalls/systemcalls.c]: @Suraj-Ajjampur will these checks ensure you return false if the `cmd` passed to `system` fails with any return code other than 0?

File: examples/systemcalls/systemcalls.c [hunk 2]
```diff
@@ -36,30 +50,94 @@ bool do_system(const char *cmd)
...
+
+    // Create a child process
+   pid_t child_pid = fork();
+
+    if (child_pid == -1) {
+        // Fork failed
+        openlog("MyProgram", LOG_PID | LOG_CONS, LOG_USER);
+        syslog(LOG_ERR, "Fork failed");
+        closelog();
+        exit(EXIT_FAILURE);
```
Comment (dwalkes) [inline examples/systemcalls/systemcalls.c]: @Suraj-Ajjampur missing `va_end` in several return paths here.

File: examples/systemcalls/systemcalls.c [hunk 3]
```diff
@@ -93,6 +167,88 @@ bool do_exec_redirect(const char *outputfile, int count, ...)
  *
 */
 
+
```
Comment (dwalkes) [inline examples/systemcalls/systemcalls.c:153]: @Suraj-Ajjampur think about how you could avoid duplicating logic between these two functions.

File: testfile.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
+home is $HOME
```
Comment (dwalkes) [inline testfile.txt:1]: @Suraj-Ajjampur no need to add this file to revision control.

**Review by dwalkes — COMMENTED** [review]
Nice work!

File: examples/systemcalls/systemcalls.c (hunk 1 above)
Comment (suraj-ajjampur) [reply examples/systemcalls/systemcalls.c]: Edited to obtain the exit status status of the child process 
```
    if (WIFEXITED(status)) {
        // Child process exited normally
        return (WEXITSTATUS(status) == 0);
    } 
    else 
    {
            // Child process did not exit normally
            return false;
    }
```

File: examples/systemcalls/systemcalls.c (hunk 3 above)
Comment (suraj-ajjampur) [reply examples/systemcalls/systemcalls.c:153]: Added Function to perform the fork execv and also check if redirection is required.

```
bool execute_command(char * command[], int redirect_fd)
{
    pid_t child_pid = fork();

    if (child_pid == -1) {
        // Fork failed
        syslog(LOG_ERR, "Fork failed");
        return false;
    }

    if (child_pid == 0) {
        // This is the child process

        // Redirect stdout if necessary
        if (redirect_fd != -1) {
            if (dup2(redirect_fd, 1) == -1) {
                syslog(LOG_ERR, "Failed to redirect stdout");
                exit(EXIT_FAILURE);
            }
        }

        // Execute the command
        if (execv(command[0], command) == -1) {
            syslog(LOG_ERR, "execv failed");
            exit(EXIT_FAILURE);
        }
    } else {
        // This is the parent process
        int status;

        // Wait for the child process to finish
        wait(&status);

        if (WIFEXITED(status)) {
            return (WEXITSTATUS(status) == 0);
        } else {
            return false;
        }
    }

    return true;  // Should never reach here
}
```

**Pull request merged**